nominal/core.py

```diff
@@ -311,6 +343,69 @@ def _from_conjure(cls, client: NominalClient, dataset: scout_catalog.EnrichedDat
         )
 
 
+@dataclass(frozen=True)
+class LogSet:
+    rid: str
+    name: str
+    timestamp_type: LogTimestampType
+    description: str | None
+    _client: NominalClient = field(repr=False)
+
+    def _stream_logs_paginated(self) -> Iterable[datasource_logset_api.Log]:
+        request = datasource_logset_api.SearchLogsRequest()
+        while True:
+            response = self._client._logset_client.search_logs(
+                self._client._auth_header,
+                log_set_rid=self.rid,
+                request=request,
+            )
+            yield from response.logs
+            if response.next_page_token is None:
+                break
+            request = datasource_logset_api.SearchLogsRequest(token=response.next_page_token)
+
+    def stream_logs(self) -> Iterable[Log]:
+        """
+        Get logs within this log set.
+        """
+        for log in self._stream_logs_paginated():
+            yield Log._from_conjure(log)
+
+    @classmethod
+    def _from_conjure(cls, client: NominalClient, log_set_metadata: datasource_logset_api.LogSetMetadata) -> Self:
+        return cls(
+            rid=log_set_metadata.rid,
+            name=log_set_metadata.name,
+            timestamp_type=_log_timestamp_type_from_conjure(log_set_metadata.timestamp_type),
+            description=log_set_metadata.description,
+            _client=client,
+        )
+
+
+@dataclass(frozen=True)
+class Log:
+    timestamp: IntegralNanosecondsUTC
+    body: str
+
+    def _to_conjure(self) -> datasource_logset_api.Log:
+        # TODO: do we need to support properties? can we leave it out on the client?
+        return datasource_logset_api.Log(
+            time=_flexible_time_to_global_conjure_api(self.timestamp),
+            body=datasource_logset_api.LogBody(
+                basic=datasource_logset_api.BasicLogBody(message=self.body, properties={}),
+            ),
+        )
+
+    @classmethod
+    def _from_conjure(cls, log: datasource_logset_api.Log) -> Self:
+        if log.body.basic is None:
+            raise ValueError("log body is empty")
```

Comment:
this is only coming from the backend - users can't change anything here, so it shouldn't be a value error. If this is a real possibility, the type should reflect it with `body: str | None`. If this shouldn't be possible then I think we should raise a `NominalError` or something to indicate that something is actually wrong with the backend, or the client-lib's understanding of the backend, if that makes sense.

Reply:
This is one of those weird situations where in Conjure we define a union type (Body) with just one option (Basic), and the generated Python stubs end up being of type Optional[Basic] - I'll change body to str | None